### PR TITLE
Added source comment ratio to report

### LINKED ISSUE #98: Include #comments, sloc, loc into complexity report
Just additional size metrics. Comments / Source code ratio could be interesting.

Idea: Extend `LLOC` class to not just increment variable `counter` but different 'counters'

### PR DESCRIPTION
**Pull request by schalkms**
This should close #98

### CODE REVIEW

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/ComplexityReport.kt [hunk 1]
```diff
@@ -24,24 +24,28 @@ class ComplexityReport : ConsoleReport() {
...
-				append("$smellPerThousandLines code smells per 1000 lloc".format(PREFIX))
-				toString()
-			}
-		} else {
-			return null
+
+		if (mcc == null) return null
+		if (lloc == null || lloc == 0) return null
+		if (sloc == null || sloc == 0) return null
+		if (cloc == null) return null
```
Comment (arturbosch) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/ComplexityReport.kt]: We will get a conflict with the ReturnCount rule, the `return null` does not look good.
Try to refactor to something like:
```kotlin
val shoudlPrintReport = when {
    mcc == null -> false
    lloc == null || lloc == 0 -> false
    ....
}

return if (shouldPrintReport) {
  // builder stuff
} else null
```

**Comment by schalkms** [comment]
The function had too much functionality. I refactored it.

**Comment by arturbosch** [comment]
I like the refactoring, good job! Make sure to use better names for ReportGenerator and Metric. They are too generic. ComplexityReportGenerator and ComplexityMetrics describe them more :D. Yeah, 100 and 1000 is always something with division or multiplication. Feel free to add them to the failfast.yml which is used to check detekt itself.

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/ComplexityReportGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	private fun cannotGenerate(): Boolean {
+		return when {
+			complexityMetric.mcc == null -> true
+			complexityMetric.lloc == null || complexityMetric.lloc == 0 -> true
+			complexityMetric.sloc == null || complexityMetric.sloc == 0 -> true
+			complexityMetric.cloc == null -> true
+			else -> {
+				numberOfSmells = complexityMetric.findings.sumBy { it.value.size }
```
Comment (arturbosch) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/ComplexityReportGenerator.kt]: Please move the calculation into a separate method and call it after the `if(cannotGenerate())`, so thte method does two different things ^^ then I will merge it :D

Comment (schalkms) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/ComplexityReportGenerator.kt]: You are right. However, the kotlin compiler does not detect the null checking before the calculation considering your approach. So I had to use the **!!** which is not what I want or to introduce separate private non nullable integers in the ComplexityReportGenerator for the complexity metrics.

Comment (arturbosch) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/ComplexityReportGenerator.kt]: Ok seems reasonable, will merge this for now, thank you!

**Comment by arturbosch** [comment]
Merged manually.

Edit: it now prints `14 comment source ratio`, should I change it to `14% comment/source ratio` ?

**Comment by schalkms** [comment]
@arturbosch Yes. Somehow it got lost. I'll fix it today.

**Pull request closed without merging**